src/storage/txn/process.rs

```diff
@@ -132,31 +132,34 @@ pub(super) fn process_read_impl<E: Engine>(
             let result = reader.scan_locks(
                 scan_key.as_ref(),
                 |lock| txn_status.contains_key(&lock.ts),
-                RESOLVE_LOCK_BATCH_SIZE,
+                RESOLVE_LOCK_BATCH_SIZE + 1, // +1 is for scan out the next lock to be scanned
             );
             statistics.add(reader.get_statistics());
-            let (kv_pairs, has_remain) = result?;
+            let (mut kv_pairs, _) = result?;
+            let has_remain = kv_pairs.len() > RESOLVE_LOCK_BATCH_SIZE;
             tls_collect_keyread_histogram_vec(tag.get_str(), kv_pairs.len() as f64);
 
             if kv_pairs.is_empty() {
                 Ok(ProcessResult::Res)
             } else {
                 let next_scan_key = if has_remain {
                     // There might be more locks.
-                    kv_pairs.last().map(|(k, _lock)| k.clone())
```

Comment:
The old way to decide `next_scan_key` would cause a dead lock now.
So I changed it.